Fix unknown output shape issue in autograph for tf.equal

yongtang opened the pull request:
This PR tries to address the issue raised in #40471 where
the output shape of an autograph consists of tf.equal
could not inference correctly. Specifically
`x.shape == [None, 10, 1]` and `y.shape == [None, 1, 4]`
only yield `shape == None` (should be `shape == [None, 10, 4]`).

The reason was that the shape inbference function for equal
didn't capture the cases where both x and y's dim are None.

This PR fixes the issue.

This PR fixes #40471.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix unknown output shape issue in autograph for tf.equal

This PR tries to address the issue raised in 40471 where
the output shape of an autograph consists of tf.equal
could not inference correctly. Specifically
`x.shape == [None, 10, 1]` and `y.shape == [None, 1, 4]`
only yield `shape == None` (should be `shape == [None, 10, 4]`).

The reason was that the shape inbference function for equal
didn't capture the cases where both x and y's dim are None.

This PR fixes the issue.

This PR fixes 40471.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

mdanatg commented on tensorflow/python/autograph/operators/logical_test.py:
```diff
@@ -83,6 +85,18 @@ def test_not_tf(self):
       t = logical.not_(self._tf_false())
       self.assertEqual(self.evaluate(t), True)
 
+  # Test case for GitHub issue 40471
```
Since the fix itself is not directly related to autograph, it would be best to have this test somewhere in a location that's closer to the `tf.equal` op.

I think `math_ops_test.cc` is a good place to verify this logic, since it already includes a basic test. Consider adding a second test which has a more complex shape, like the ones tested here. Feel free to add a new test case (e.g. `TEST(MathOpsTest, EqualOp)`).

Optionally, there is also `tensorflow/python/kernel_tests`, which could be expanded by adding an `equal_test.py` to contain a few more test cases, which should include:
 * matching shapes
 * incompatible shapes
 * dynamic, matching shapes
 * dynamic, incompatible shapes (this last one is most important to make sure that op kernels weren't relying on the shape inference to behave correctly)

yongtang committed:
Move tf.equal shape inference test to math_ops_test.cc

also added additional shape inference test cases

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@mdanatg Thanks for the review. I have updated the PR to move the test case  to math_ops_test.cc. Also added additional shape inference cases with a better coverage. Please take a look and let me know if there are any issues.

mdanatg commented on tensorflow/core/ops/math_ops_test.cc:
```diff
@@ -604,4 +604,22 @@ TEST(MathOpsTest, SobolSample) {
+
+TEST(MathOpsTest, EqualOp) {
+  ShapeInferenceTestOp op("Equal");
+  AddNodeAttr("incompatible_shape_error", true, &op.node_def);
```
(Optional) This attribute should not be necessary - do tests fail without it?

yongtang replied:
@mdanatg `incompatible_shape_error` Attr is part of the `Equal` op due to the following line, so it is needed.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/math_ops.cc#L718

mdanatg reviewed (approved):
Thank you!
